File: Graphs/Python/AdjacencyList/HashMapsBased/Graph.py
import os
import sys
from Vertex import Vertex
from Edge import Edge
from HashTable import HashTable
from PQueue import PQueue

class Graph:
    
    def __init__(self,size):
        self.size = size
        self.order = 0
        self.degree = 0
        self.e = size*[None] 

        for i in range(size):
            self.e[i] = Edge()
            self.e[i].v = None
    
    def printGraph(self):
        if self.order == 0:
            print("The graph is empty \n")
        else:
            for i in range(self.size):
                if self.e[i].v != None:
                    print("This is vertex: ",self.e[i].v.label)
                    aux = self.e[i].v.next
                    if aux == None:
                        print("\n This vertex has no connections \n")
                    else:
                        print(" and it has the following connections: \n")
                        while aux!= None:
                            print("=========>",aux.label) 
                            aux = aux.next

                        print("\n \n")
    
    def addVertex(self,e:Vertex,t:HashTable):
        if(self.order >= self.size):
            printf("There is no space where to add a vertex")
        else:
            self.e[self.order].v = e
            t.insertHT(e.label,self.order)
            self.order +=1
    
    def addEdge(self,src,dst,e_w,t:HashTable):
        src_id = t.searchHT(src)
        dst_id = t.searchHT(dst)

        if src_id == None or dst_id == None:
            print("Either the source or destination vertex does not exist")
        else:
            temp = Vertex(self.e[dst_id.value].v.label,self.e[dst_id.value].v.id,self.e[dst_id.value].v.data)
            temp.edge_weight = e_w
            aux = self.e[src_id.value].v
            while aux.next != None:
                aux = aux.next
            
            aux.next = temp
            self.degree+=1

    def deleteVertex(self,src,t:HashTable):
        src_v = t.searchHT(src)

        if src_v == None:
            print("This vertex does not exist \n")
        else:
            for i in range(self.size):
                aux = self.e[i].v
                prev = None
                b = False

                while aux != None and not(b):
                    if aux.label == src:
                        b = True
                    else:
                        prev = aux
                        aux = aux.next
                
                if b:
                    if prev == None:
                        self.e[i].v = None
                    else:
                        prev.next = aux.next

            self.order-=1
            t.deleteHT(src)

    
    def deleteEdge(self,src,dst,t:HashTable):
        src_id = t.searchHT(src)
        dst_id = t.searchHT(dst)

        if src_id == None or dst_id == None:
            print("either the source or destination vertex does not exist \n")
        else:
            aux = self.e[src_id.value].v.next
            prev = None
            b = False

            while aux != None and not(b):
                if aux.label == dst:
                    b = True
                else:
                    prev = aux
                    aux = aux.next
            
            if b:
                self.degree-=1
                if prev == None:
                    self.e[src_id.value].v.next = self.e[src_id.value].v.next.next
                else:
                    prev.next = aux.next
            else:
                print("There is no connection between ",src," and ",dst)
                print()

    def dijkstraAlgo(self,src,t:HashTable):
        v_src = t.searchHT(src)

        if v_src == None:
            print("This vertex does not exist \n")
        else:
            INFINITY = 10**18
            res = self.size*[None]
            q = PQueue(self.size)
            
            for i in range(self.size):
                temp = Vertex(self.e[i].v.label,self.e[i].v.id,self.e[i].v.data)
                if src != self.e[i].v.label:
                    temp.edge_weight = INFINITY
                    res[i] = temp
                else:
                    temp.edge_weight = 0
                    res[i] = temp
                    q.insert(temp)
            
            while q.num_vd > 0:
                temp = q.pop()
                temp_id = t.searchHT(temp.label)
                aux = self.e[temp_id.value].v.next
                
                while aux != None:
                    total_w = temp.edge_weight + aux.edge_weight
                    aux_id = t.searchHT(aux.label)

                    if total_w < res[aux_id.value].edge_weight:
                        res[aux_id.value].edge_weight = total_w
                        q.insert(res[aux_id.value]) 

                    aux = aux.next
        
            return res









"""Testing"""

print("Give me the size of your graph: \n")
N = int(input())
g = Graph(N)
t = HashTable(N)

i = 0
while i < N:
    print("Give me the label, id and data of vertex: ",i,"\n")
    e_l = input()
    e_id = int(input())
    e_d = input()

    temp = Vertex(e_l,e_id,e_d)
    g.addVertex(temp,t)
    i+=1
    print()

print()

b = 1
while b:
    print("Would you like to add edges 1 for yes\0 for no \n")
    b = int(input())

    if b:
        print("Give me the source and destination vertices as well as the edge weight: \n")
        v_src = input()
        v_dst = input()
        v_w = int(input())
        g.addEdge(v_src,v_dst,v_w,t)

print()

g.printGraph()

while True:
    print("Give me the source vertex for which you would to calculate the short distance to all other vertices:")
    src = input()
    r = g.dijkstraAlgo(src,t)
    print()
    
    if r != None:
        print("The shortest distance from ",src," to each vertex is: \n")
        print()
        for x in range(len(r)):
            print("==========> ",r[x].label,": ",r[x].edge_weight," \n")
    
    print()
    

File: Graphs/C++/IDBased/Graph.cpp
#include "Vertex.h"
#include "Edge.h"
#include "PQueue.h"

#include<iostream>
#include<malloc.h>
#include<stdlib.h>
#include<vector>



using std::cin;
using std::cout;
using std::end;
using std::vector;

template <typename T1,typename T2,typename T3>
class Graph{

        public:
              
              int size;
              int num_vertices;
              int degree;
              vector<Edge<T1,T2,T3>> e;

              Graph(int size):size(size),num_vertices(0),e(size),degree(0){}

              void printGraph(Graph<T1,T2,T3>& gr){

                if(gr.num_vertices == 0)
                cout << "The graph is empty";
                else{
                   for(int i = 0;i<gr.size;i++){
                    if(gr.e[i].v != nullptr)
                    {
                      cout << "This is vertex: " << gr.e[i].v->label << " \n";
                      Vertex<T1,T2,T3> *aux = gr.e[i].v;
                      if(aux->next == nullptr)
                      cout << "This vertex has no connections \n \n";
                      else{
                        cout << "This vertex" << aux->label << " which has the following connections: \n";
                        while(aux != nullptr){
                          cout << "============> " << aux->label << "\n";
                          aux = aux->next;
                        }
                      }
                    }
                   }    
               }
              }


              void insertVertex(Graph<T1,T2,T3>& gr,Vertex<T1,T2,T3> *vx){
                if(gr.num_vertices >= gr.size)
                cout << "There are no spots where to insert a vertex \n";
                else{
                  if(gr.e[vx->id].v == nullptr){
                    gr.e[vx->id].v = vx;
                    gr.num_vertices++;
                  }else cout << "This vertex position is alread taken";            
                }
              }

              void insertEdge(Graph<T1,T2,T3> gr,T1 id1, T2 id2){
                    if(gr.e[id1].v == nullptr || gr.e[id2].v == nullptr)
                    cout << "One of the vertices does not exist \n";
                    else{
                      Vertex<T1,T2,T3> *aux = gr.ed[id1].v;
                      while(aux->next != nullptr)
                      aux = aux->next;

                      Vertex<T1,T2,T3> *vx = new Vertex<T1,T2,T3>(gr.ed[2].v.id,gr.ed[2].v.label,gr.ed[2].v.data);
                      aux->next = vx;
                      gr.degree++;
                    }
              }

              void deleteVertex(Graph<T1,T2,T3> gr,T1 id){
                if(gr.e[id].v == nullptr)
                cout << "This vertex does not exist";
                else{
                  for(int i = 0;i<gr.size;i++){
                    if(gr.e[i].v != nullptr){
                      Vertex<T1,T2,T3> *aux = gr.e[i].v;
                      Vertex<T1,T2,T3> *prev = nullptr;

                      while(aux != nullptr && aux->id != id){
                        prev = aux;
                        aux = aux->next;
                      }

                      if(aux != nullptr){
                        if(prev == nullptr)
                        gr.e[i].v =gr.e[i].v->next;
                        else prev->next = aux->next;
                      }

                    }
                  }
                }
              }

               void deleteEdge(Graph<T1,T2,T3> gr,T1 id1, T2 id2){
                if(gr.e[id1].v == nullptr || gr.e[id2].v == nullptr)
                cout << "either Vertex" << id1 << " or Vertex " << id2 << " does not exist \n";
                else{
                  Vertex<T1,T2,T3> *aux = gr.e[id1].v->next;
                  Vertex<T1,T2,T3> *prev = nullptr;
                  bool b = false;

                  while(aux !=nullptr && !b){
                    if(aux.id == id2)
                    b = true;
                    else{
                      prev = aux;
                      aux = aux->next;
                    }
                  }

                  if(!b)
                  cout << "There is no connection between vertex " << id1 << " and Vertex " << id2 << " \n";
                  else{
                    if(prev == nullptr)
                    gr.e[id1].v->next = gr.e[id1].v->next->next;
                    else prev->next = aux->next;
                  } 

                }
              }

};


int main(){
    
    int N;
    cout << "Give me the size of your graph\n";
    cin >> N;

    Graph<int,string,string> g(N);

    int numv;
    cout << "Give me the number of vertices you would like to insert in your graph \n";
    cin >> numv;
    cout << "\n";

    int i = 0;
    int v_id;
    string v_l, v_d;

    while(i<numv){
      cout << "Give me the id, label and data of vertex " << i << " \n";
      cin >> v_id;
      cin >> v_l;
      cin >> v_d;
      Vertex<int,string,string> *ve = new Vertex<int,string,string>(v_id,v_l,v_d);
      g.insertVertex(g,ve);
      i++; 
      cout << "\n";
    }

   g.printGraph(g);

   cout << "\n";

    exit(1);
}